src/lib/registration/sw_mgmt.rb

```diff
@@ -655,23 +655,31 @@ def self.init_target(destdir)
         # will use the old value from the upgraded system which might not
         # match the new target_distro from the media and might result in ignoring
         # service repositories (bsc#1094865)
-        options = { "target_distro" => target_distribution }
+        options = { "target_distro" => target_distribution(destdir) }
         Pkg.TargetInitializeOptions(destdir, options)
       else
         Pkg.TargetInitialize(destdir)
       end
     end
 
     # get the target distribution for the new base product
+    # @param destdir [String] the target directory
     # @return [String] target distribution name or empty string if not found
-    def self.target_distribution
+    def self.target_distribution(destdir)
+      # ensure the target is initialized
+      Pkg.TargetInitialize(destdir)
+      # the sources are initialized by the Product.FindBaseProducts call internally
       base_products = Product.FindBaseProducts
 
       # empty target distribution disables service compatibility check in case
       # the base product cannot be found
       target_distro = base_products ? base_products.first["register_target"] : ""
       log.info "Base product target distribution: #{target_distro.inspect}"
 
+      # close both target and sources to fully reinitialize later
+      Pkg.SourceFinishAll
+      Pkg.TargetFinish
```

Comment:
what is performance impact? Does initialize also download and parse all repository metadata?

Reply:
There is no performance impact, libzypp has a 10 minutes timeout by default (`repo.refresh.delay = 10`). The repository metadata are not loaded (not even checked for a change on the server) if they were downloaded during the last 10 minutes.

See https://github.com/openSUSE/libzypp/blob/master/zypp.conf#L121-L137

So registering against SCC and then immediately restarting the package manager has no performance impact.